lib/Sema/CSSimplify.cpp

```diff
@@ -757,22 +757,37 @@ class ArgumentFailureTracker : public MatchCallArgumentListener {
 // Match the argument of a call to the parameter.
 ConstraintSystem::TypeMatchResult
 constraints::matchCallArguments(ConstraintSystem &cs, bool isOperator,
-                                Type argType, Type paramType,
+                                ArrayRef<AnyFunctionType::Param> args,
+                                ArrayRef<AnyFunctionType::Param> params,
                                 ConstraintLocatorBuilder locator) {
+  // Extract the parameters.
+  ValueDecl *callee;
+  unsigned calleeLevel;
+  ArrayRef<Identifier> argLabels;
+  SmallVector<Identifier, 2> argLabelsScratch;
+  bool hasTrailingClosure = false;
+  std::tie(callee, calleeLevel, argLabels, hasTrailingClosure) =
+    getCalleeDeclAndArgs(cs, locator, argLabelsScratch);
 
-  if (paramType->isAny()) {
-    if (argType->is<InOutType>())
-      return cs.getTypeMatchFailure(locator);
+  llvm::SmallBitVector defaultMap =
+    computeDefaultMap(params, callee, calleeLevel);
 
-    // If the param type is Any, the function can only have one argument.
-    // Check if exactly one argument was passed to this function, otherwise
-    // we obviously have a mismatch.
-    if (auto tupleArgType = dyn_cast<TupleType>(argType.getPointer())) {
-      if (tupleArgType->getNumElements() != 1 ||
-          tupleArgType->getElement(0).hasName()) {
-        return cs.getTypeMatchFailure(locator);
-      }
-    }
+  // Apply labels to arguments.
+  SmallVector<AnyFunctionType::Param, 8> argsWithLabels;
+  argsWithLabels.append(args.begin(), args.end());
+  AnyFunctionType::relabelParams(argsWithLabels, argLabels);
```

Comment:
I think this requires a bit of clarification... Do `args` not always have labels attached? I know that the same thing was done by `decomposeArgType` before, but maybe - since this code is being refactored - it's a good time to add some clarify here.

Reply:
Yeah I didn’t understand that part. Maybe @DougGregor remembers the historical context here. I would expect that since arguments come from a TupleExpr they always have labels attached. Doug, do you remember why decomposeArTypes() would introduce labels?

Reply:
I don't remember off-hand, but maybe it is because the solver might have adjusted for incorrect labels without updating the tuple type?